### DISCUSSION TITLE
Running http/mqtt server

### DISCUSSION THREAD

**Discussion by gawelkot** [cesanta/mongoose#2175] [Ask questions, get answers] [answered]
Hi,
I am trying to get http and mqtt server running but I have problem with getting these 2 protocols working. 

I mixed http and mqtt server examples. I am using 1 event manager and 2 event handler functions (1 for protocol). When I comment 1 of these listen functions the other protocol is working without any problems. When I use both protocols only MQTT is working. I see ERR_SOCKET_NOT_CONNECTED error in my browser after http connection.

Can you guide me what the problem can be? I am working on STM32H7 Nucleo board with freertos. This is how my functions look like:

```
static const char *s_listen_http= "http://0.0.0.0:80";
static const char *s_listen_mqtt= "mqtt://0.0.0.0:1883";

...

//MQTT handler
static void mqtt_handler(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
  if (ev == MG_EV_MQTT_CMD) {
  ...
  }
}

//HTTP handler
static void http_handler(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
  if (ev == MG_EV_HTTP_MSG) {
  ...
  }
}

void server(void *args) {
  mg_mgr_init(&mgr);
  MG_INFO( ("Starting Mongoose v%s", MG_VERSION)); // Tell the world
  void *a = mg_mqtt_listen(&mgr, s_listen_mqtt, mqtt_handler, NULL); //MQTT listener
  void *b = mg_http_listen(&mgr, s_listen_http, http_handler, &mgr); // Web listener

  while (1) {
    mg_mgr_poll(&mgr, 50); // Infinite event loop
  }

  mg_mgr_free(&mgr); // Unreachable
}
```

**Comment by scaprile** [comment, marked as answer]
https://mongoose.ws/documentation/tutorials/http-server/
https://mongoose.ws/documentation/tutorials/mqtt-server/

> I see ERR_SOCKET_NOT_CONNECTED error in my browser after http connection.

Unfortunately "your browser" and its messages are out of reach for me. What do you see in:
- Mongoose log
- Wireshark
?
> I am working on STM32H7 Nucleo board with freertos.

Whose port and configuration ? Cube ?
What is your stack size for this task ?
What TCP/IP stack are you using ? Whose port and configuration ? Cube's lwIP ? CMSIS_RTOS v1 ? Number of connections allowed ? If you answered yes to all then add MEMP_NUM_NETCONN=10 or more to your preprocessor symbols
https://mongoose.ws/documentation/tutorials/stm32/nucleo-f746zg-cube-freertos-lwip/images/netconn.webp
And read this tutorial:
https://mongoose.ws/documentation/tutorials/stm32/nucleo-f746zg-cube-freertos-lwip/

PS: to see Mongoose log, even though the tutorial above will link you to it: https://mongoose.ws/documentation/tutorials/stm32/nucleo-f746zg-cube/#step-3-redirect-printf-output-to-usart3 (check the correct USART connected to the ST-Link in your board)

**Reply by gawelkot** [reply]
Thank you!
All the answers were yes so your sugested sollution with MEMP_NUM_NETCONN parameter worked well. I missed this part of LWIP tutorial as it is not included in tutorial for STM32H7 Nucleo board with Freertos + lwip.

**Reply by scaprile** [reply]
Yeah, we are slowly making sure all tutorials carry the same vital info, I'll mark that one for prompt review. Thanks.

**Reply by noName2005** [reply]
I would like to achieve a similar setup as the original poster (OP) but using an ESP32-WROOM-32 board. My goal is to have it function as an MQTT broker while simultaneously serving a local server. Are there any specific considerations or steps I should be mindful of to make it work?

**Reply by scaprile** [reply]
Please don't thread hijack.
The answer is no, please follow the documentation
